lib/chai/core/assertions.js

```diff
@@ -528,10 +528,25 @@ module.exports = function (chai, _) {
    */
 
   Assertion.addProperty('empty', function () {
-    var obj = flag(this, 'object');
-    new Assertion(obj).to.exist;
+    var val = flag(this, 'object')
+      , itemsCount;
+
+    switch (_.type(val)) {
+      case 'array':
+      case 'string':
+        itemsCount = val.length;
+        break;
+      case 'function':
```

Comment:
This should handle generators too (because of recent updates to `type-detect`) and `async` function in the future.